### ISSUE TITLE
Bug: [no-unused-vars] Generic passed to class reference during class property assignment not picked up as being used.

### ISSUE THREAD

**Issue by stagas** [typescript-eslint/typescript-eslint#5317]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.7.4&sourceType=module&code=MYGwhgzhAEBiD28A8AVAfNA3gKGtAbmCAK4CmAXNCtgL7baiQwBCYATqhjngGaLQBeOIk616AO1IB3aKzYAKAJTYgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgwBtSImjQO0SAF1wYAL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
class Foo<T> {
  value: T
}

class Bar<T> {
  foo = Foo<T>
}

new Bar()
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unused-vars": ["error"]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

To pick up that T in `foo = Foo<T>` is being used.

### Actual Result

`'T' is defined but never used. 5:11 - 5:12`

### Additional Info

The ability to modify function references with generics is a new feature in TS 4.7, so this is probably the reason the rule hasn't updated yet to support it.

### Versions

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `5.30.5` |
| `@typescript-eslint/parser` | `5.30.5` |
| `TypeScript` | `4.7.4` |
| `ESLint` | `8.15.0` |
| `node` | `web` |
